libyul/optimiser/VariableAssignmentMap.h

```diff
@@ -0,0 +1,85 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#pragma once
+
+#include <libyul/YulString.h>
+
+#include <set>
+#include <unordered_map>
+
+namespace solidity::yul
+{
+
+/**
+ * Class that implements a reverse lookup for an ``unordered_map<YulString, set<YulString>>`` by wrapping the
+ * two such maps - one ordered, and one reversed, e.g.
+ *
+ *  m_ordered           m_reversed
```

Comment:
It's ordered in the sense that it's the opposite of reversed. `m_assigments` and `m_uses` isn't really correct either, e.g. if we have 
```
a = x + y;
```
then `(a,x)` and `(a,y)` could be assignments - but what are the uses? `(x, a)`, (y, a)`? That doesn't really make sense to me either. `lvalues` and `rvalues` do make sense, but are then somewhat confusing in terms of C++ semantics. In any case, I've spend quite a while trying to come up with names for these, and I'm still convinced these are the best, especially since the whole purpose of this PR is to implement a reverse lookup.